pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go

```diff
@@ -0,0 +1,616 @@
+
+package core
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+)
+
+type GroupQuotaManager struct {
+	// Lock used for resourceKeys/quotaInfoMap/quotaTreeWrapper change
+	Lock                sync.RWMutex
+	totalResourceIndeed v1.ResourceList // totalResource without systemQuotaGroup and DefaultQuotaGroup's used Quota
+	totalResource       v1.ResourceList // totalResource with systemQuotaGroup and DefaultQuotaGroup's used Quota
+	resourceKeys        map[v1.ResourceName]struct{}
+	// quotaInfoMap stores all the nodes, it can help get all parents conveniently
+	quotaInfoMap map[string]*QuotaInfo
+	// quotaTreeWrapperMap helps calculate the subGroups' runtimeQuota in one quotaGroup
+	quotaTreeWrapperMap map[string]*QuotaTreeWrapper
+	// quotaTopoNode only stores the topology of the quota
+	quotaTopoNode                    map[string]*QuotaTopoNode
+	scaleMinQuotaEnabled             bool
+	scaleMinQuotaWhenOverRootResInfo *ScaleMinQuotaWhenOverRootResInfo
+	once                             sync.Once
+}
+
+func NewGroupQuotaManager(systemGroupMax, defaultGroupMax v1.ResourceList) *GroupQuotaManager {
+	quotaManager := &GroupQuotaManager{
+		totalResourceIndeed:              v1.ResourceList{},
+		totalResource:                    v1.ResourceList{},
+		resourceKeys:                     make(map[v1.ResourceName]struct{}),
+		quotaInfoMap:                     make(map[string]*QuotaInfo),
+		quotaTreeWrapperMap:              make(map[string]*QuotaTreeWrapper),
+		quotaTopoNode:                    make(map[string]*QuotaTopoNode),
+		scaleMinQuotaWhenOverRootResInfo: NewScaleMinQuotaWhenOverRootResInfo(),
+	}
+	quotaManager.quotaInfoMap[extension.SystemQuotaName] = NewQuotaInfo(true, true, extension.SystemQuotaName, extension.SystemQuotaName)
+	quotaManager.quotaInfoMap[extension.SystemQuotaName].setMaxQuotaNoLock(systemGroupMax)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName] = NewQuotaInfo(true, true, extension.DefaultQuotaName, extension.DefaultQuotaName)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName].setMaxQuotaNoLock(defaultGroupMax)
+	quotaManager.quotaTreeWrapperMap[extension.RootQuotaName] = NewQuotaTreeWrapper(extension.RootQuotaName)
+	return quotaManager
+}
+
+func (gqm *GroupQuotaManager) SetScaleMinQuotaEnabled(flag bool) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.scaleMinQuotaEnabled = flag
+	klog.V(3).Infof("Set ScaleMinQuotaEnabled, flag:%v", gqm.scaleMinQuotaEnabled)
+}
+
+func (gqm *GroupQuotaManager) UpdateClusterTotalResource(deltaRes v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	klog.V(3).Infof("UpdateClusterResource deltaRes:%v", deltaRes)
+	gqm.updateClusterTotalResourceNoLock(deltaRes)
+}
+
+func (gqm *GroupQuotaManager) updateClusterTotalResourceNoLock(deltaRes v1.ResourceList) {
```

Comment:
I don't this updateClusterTotalResourceNoLock is neccessory for "UpdateGroupDeltaUsed" should be killed.